kernel/rtlil.cc

```diff
@@ -1747,6 +1747,12 @@ namespace {
 				return;
 			}
 
+			if (cell->type == ID($scopeinfo)) {
+				param(ID::TYPE);
```

Comment:
I wonder if the type shouldn't be an attribute as well, but I don't have much of an argument in favor of it right now.

Reply:
I would like to not have to worry about invalid types in the debugger frontend, meaning it should be verified in `check()`. That requires it to be a parameter.

Reply:
If I'm not mistaken, we're so far not using attributes for anything that's unconditionally required to be present on a cell, which was why I agreed with @whitequark's suggestion on slack to use a parameter here.

I'm not sure if we should limit the available scope types to a fixed set, leave it open to allow arbitrary user defined scope types, or a hybrid where e.g. user defined scope types share a prefix identifying them as such (similar to private and public names).

In the current PR I only check for presence of the parameter, as I wanted to discuss this first.

Reply:
The scope types are a contract between the compiler and the debugger. The debugger must understand every scope type to properly present them in the interface, so the list of scope types should be closed (with potential for custom ones). This is how DWARF works for example.

Reply:
Do you already have a list of scope types besides "module" that should be initially allowed?

I'm leaning towards having the list of scope types be completely closed initially, i.e. deferring how we handle custom scope types until an actual need for them arises. I can think of several ways, but wouldn't be able to make a good decision without a concrete use case in mind.

Reply:
I would say `module` and `struct`.